### PR TITLE
Allow full locale in .arb files

### PR DESCRIPTION
**Pull request by TahaTesser**
Reland https://github.com/flutter/flutter/pull/93401 with fixes added for library imports and fix for languages with script code not being generated

Also updated the test to include an arb file with script code.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: packages/flutter_tools/lib/src/localizations/gen_l10n.dart [hunk 1]
```diff
@@ -596,7 +596,7 @@ String _generateLookupBody(
       switchClauseDeferredLoadingTemplate : switchClauseTemplate)
       .replaceAll('@(localeClass)', '$className${locale.camelCase()}')
       .replaceAll('@(appClass)', className)
-      .replaceAll('@(library)', '${fileName}_${locale.languageCode}');
+      .replaceAll('@(library)', '${fileName}_${locale.toString()}');
```
Comment (TahaTesser) [inline packages/flutter_tools/lib/src/localizations/gen_l10n.dart:599]: @HansMuller 
Fixed this where import was only generated with language code, it should be full locale.

File: packages/flutter_tools/lib/src/localizations/gen_l10n.dart [hunk 2]
```diff
@@ -1150,9 +1150,7 @@ class LocalizationsGenerator {
   // itself does not generate the output files.
   String _generateCode() {
     bool isBaseClassLocale(LocaleInfo locale, String language) {
-      return locale.languageCode == language
-          && locale.countryCode == null
-          && locale.scriptCode == null;
+      return locale.languageCode == language;
```
Comment (TahaTesser) [inline packages/flutter_tools/lib/src/localizations/gen_l10n.dart:1153]: @HansMuller 
Languages with script code also needed a base locale file so which we won't anymore. This passes all Flutter Gallery tests locally.

File: packages/flutter_tools/lib/src/localizations/gen_l10n.dart (hunk 1 above)
Comment (HansMuller) [reply packages/flutter_tools/lib/src/localizations/gen_l10n.dart:599]: The `toString()` shouldn't be needed here

File: packages/flutter_tools/lib/src/localizations/gen_l10n.dart (hunk 2 above)
Comment (HansMuller) [reply packages/flutter_tools/lib/src/localizations/gen_l10n.dart:1153]: This method doesn't appear to mean that its locale is the "base" locale anymore.

What does this line, from `getLocalesForLanguage()` mean now?
```dart
          return !isBaseClassLocale(locale, language) && locale.languageCode == language;
```

**Review by HansMuller — COMMENTED** [review]
I'm afraid this feels a bit like hacking.

File: packages/flutter_tools/lib/src/localizations/gen_l10n.dart (hunk 1 above)
Comment (TahaTesser) [reply packages/flutter_tools/lib/src/localizations/gen_l10n.dart]: Thanks, fixed

File: packages/flutter_tools/lib/src/localizations/gen_l10n.dart (hunk 2 above)
Comment (TahaTesser) [reply packages/flutter_tools/lib/src/localizations/gen_l10n.dart:1127]: That's a good point, that was an oversight from me.  I updated the code just check for locale language

**Comment by TahaTesser** [comment]
@HansMuller 
I have updated the code to code base, remove unnecessary conditions, remove subclass functions which are no longer needed, update template class name and function names to match new behavior

**Review by HansMuller — COMMENTED** [review]
There's something wrong here. It appears to be a "pre-existing condition" rather than a problem with this PR, however this PR is going to make it worse. 

I believe the original motivation for having a base class based on the language and then subclasses for each regional variation, was to minimize the amount of code generated - since most regional localizations only override a subset of the translations. It was assumed that the ".arb" files with regional localizations would only contain the overrides. An important counter-example is the [Flutter Gallery](https://github.com/flutter/gallery/blob/master/lib/l10n), where we appear to be generating mountains of code unnecessarily.

The problem with this approach, highlighted in https://github.com/flutter/flutter/issues/92731, is that sometimes developers would prefer to specify only regional localizations. Presumably, this is because they're all different, so there's nothing to be gained by having a base class and smaller subclasses that override a few translations from the base class. 

Will follow-up with @jpnurmi about this on #92731

**Comment by TahaTesser** [comment]
@HansMuller 
Here tried to come up with a new solution. 
Now only locales without base class generate separate (since they don't have base class, they can't be inherited like a locale where the base class is provided)
Here `AppLocalizationsEnCa` depends on base class `AppLocalizationsEn`
ex:
`app_localizations_en.dart`
```dart
import 'app_localizations.dart';

/// The translations for English (`en`).
class AppLocalizationsEn extends AppLocalizations {
  AppLocalizationsEn([String locale = 'en']) : super(locale);

  @override
  String get helloWorld => 'Hello World!';
}

/// The translations for English, as used in Canada (`en_CA`).
class AppLocalizationsEnCa extends AppLocalizationsEn {
  AppLocalizationsEnCa(): super('en_CA');

  @override
  String get helloWorld => 'Hello World!';
}

/// The translations for English, as used in New Zealand (`en_NZ`).
class AppLocalizationsEnNz extends AppLocalizationsEn {
  AppLocalizationsEnNz(): super('en_NZ');

  @override
  String get helloWorld => 'Hello World!';
}

```

Similar a french locale without a base class can't have inherited class
So I'm generating it in a separate class
`app_fr.arb` (this PR no longer forces to match arb file name and @@locale as requested in the issue.
```console
{
  "@@locale": "fr_FR",
  "helloWorld": "Bonjour le monde!"
}

```
`app_localizations_fr_FR.dart`(`fr_FR` locale is generated without a base class based on `@@locale`)
```dart
import 'app_localizations.dart';

/// The translations for French, as used in France (`fr_FR`).
class AppLocalizationsFrFr extends AppLocalizations {
  AppLocalizationsFrFr([String locale = 'fr_FR']) : super(locale);

  @override
  String get helloWorld => 'Bonjour le monde!';
}
```
While you `en` wich has a  base class, it is still generated in a single file

Let me know if you like this solution better, thanks

**Comment by christopherfujino** [comment]
friendly ping @HansMuller

**Comment by HansMuller** [comment]
https://github.com/flutter/flutter/issues/92731#issuecomment-995349103

**Comment by christopherfujino** [comment]
@TahaTesser is this ready to be re-reviewed?

**Comment by christopherfujino** [comment]
@TahaTesser I'm marking this as a draft as it looks like it's blocked, so that it doesn't show up in the tool PR triage meeting. Feel free to mark it ready to review when it is!

**Pull request closed without merging**